src/Build/BackEnd/Node/RarNode.cs

```diff
@@ -0,0 +1,99 @@
+
+using System;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Internal;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Server;
+
+namespace Microsoft.Build.Execution
+{
+    public sealed class RarNode : INode
+    {
+        /// <summary>
+        /// The amount of time to wait for the client to connect to the host.
+        /// </summary>
+        private const int ClientConnectTimeout = 60000;
+
+        public NodeEngineShutdownReason Run(bool nodeReuse, bool lowPriority, out Exception shutdownException)
+        {
+            shutdownException = null;
+            using CancellationTokenSource cts = new CancellationTokenSource();
+            string pipeName = CommunicationsUtilities.GetRarPipeName(nodeReuse, lowPriority);
+            RarController controller = new RarController(pipeName, NamedPipeUtil.CreateNamedPipeServer);
+
+            Task<int> rarTask = controller.StartAsync(cts.Token);
+
+            Handshake handshake = NodeProviderOutOfProc.GetHandshake(enableNodeReuse: nodeReuse,
+                                                                     enableLowPriority: lowPriority, specialNode: true);
+            Task<NodeEngineShutdownReason> msBuildShutdown = RunShutdownCheckAsync(handshake, cts.Token);
+
+            int index = Task.WaitAny(msBuildShutdown, rarTask);
+            cts.Cancel();
+
+            if (index == 0)
+            {
+                // We know that this task is completed so we can get Result without worring about waiting for it
+                return msBuildShutdown.Result;
+            }
+            else
+            {
+                // RAR task can only exit with connection error or timeout
+                return NodeEngineShutdownReason.ConnectionFailed;
+            }
+        }
+
+        public NodeEngineShutdownReason Run(out Exception shutdownException)
+        {
+            return Run(false, false, out shutdownException);
+        }
+
+        private async Task<NodeEngineShutdownReason> RunShutdownCheckAsync(Handshake handshake, CancellationToken cancellationToken = default)
+        {
+            string pipeName = NamedPipeUtil.GetPipeNameOrPath("MSBuild" + Process.GetCurrentProcess().Id);
+
+            while (true)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                    return NodeEngineShutdownReason.Error;
+
+                using NamedPipeServerStream serverStream = NamedPipeUtil.CreateNamedPipeServer(pipeName, maxNumberOfServerInstances: NamedPipeServerStream.MaxAllowedServerInstances);
+                await serverStream.WaitForConnectionAsync(cancellationToken).ConfigureAwait(false);
+
+                bool connected = NamedPipeUtil.ValidateHandshake(handshake, serverStream, ClientConnectTimeout);
```

Comment:
Is there any way to use the existing readloop infastructure to do this? Or is that unique per node type anyway so reimplementing it here is just How It's Done?

Reply:
InProc and OutOfProc have each own endpoint. If it was just simple reuse of code I would go for that, but it didn't look like that. So I went this quite straightforward way.